Fix Demo modules

MironAtHome opened the pull request:
Changes:
subfolder callback, module cheese
  Fix __declspec(dllexport) function decorator for the function find_cheeses
  Modified cheesefinder.h and cheesefinder.c
subfolder libraries, module call_mymath
Fixes
1 same function decorator
  Modified mymath.h and mymath.c
2 add #include directive for the file "mymath.c"
  change #include directive for the file "math.h" to angle brackets to ensure proper C standard file included
3 modify file Setup.py
  add check not sys.platform.startswith("win") prior call to gcc compiler
  make call to cythonize conditional upon check for library file presence
  add second variant of cythonize for windows without libraries and library_dirs parameters
Modify Setup.py under Demos subfolder
1 add to  the list of exclusions callback and libraries subfolders and their respective modules
2 organize subset of custom extensions into array
3 ensure include_dirs for the extensions point to correct folders
4 ensure sources are correctly provided
Link issue https://github.com/cython/cython/issues/7233

MironAtHome committed:
Fix Demo modules
subfolder callback, module cheese
Fix __declspec(dllexport) function decorator for the functionn find_cheeses
Modified cheesefinder.h and cheesefinder.c
subfolder libraries, module call_mymath
Fix
1 same function decorator
  Modifled mymath.h and mymath.c
2 file name in the #include directive
  file mymath.c
3 mofify file Setup.py
  add check not sys.platform.startswith("win") prior call to gcc compiler
  make call to cythonize conditional upon check for library file presence
  add second variant of cythonize for windows without libraries and
  library_dirs parameters
Modify Setup.py under Demos subfolder
1 add to  the list of exclusions callback and libraries
2 organize custom extensions into array
3 ensure include_dirs for the extensions point to correct folders
4 ensure sources are correctly provided

MironAtHome committed:
Fix Demo modules
subfolder callback, module cheese
Fix __declspec(dllexport) function decorator for the functionn find_cheeses
Modified cheesefinder.h and cheesefinder.c
subfolder libraries, module call_mymath
Fix
1 same function decorator
  Modifled mymath.h and mymath.c
2 add #include directive for the file "mymath.c"
  change #include directove for the file "math.h" to angle brackets
  to ensure proper C standard file included
3 mofify file Setup.py
  add check not sys.platform.startswith("win") prior call to gcc compiler
  make call to cythonize conditional upon check for library file presence
  add second variant of cythonize for windows without libraries and
  library_dirs parameters
Modify Setup.py under Demos subfolder
1 add to  the list of exclusions callback and libraries
2 organize custom extensions into array
3 ensure include_dirs for the extensions point to correct folders
4 ensure sources are correctly provided

MironAtHome committed:
Fix Demo modules
subfolder callback, module cheese
Fix __declspec(dllexport) function decorator for the functionn
find_cheeses
Modified cheesefinder.h and cheesefinder.c
subfolder libraries, module call_mymath
Fix
1 same function decorator
  Modifled mymath.h and mymath.c
2 add #include directive for the file "mymath.c"
  change #include directove for the file "math.h" to angle brackets
  to ensure proper C standard file included
3 mofify file Setup.py
  add check not sys.platform.startswith("win") prior call to gcc compiler
  make call to cythonize conditional upon check for library file
  presence
  add second variant of cythonize for windows without libraries and
  library_dirs parameters
Modify Setup.py under Demos subfolder
1 add to  the list of exclusions callback and libraries
2 organize custom extensions into array
3 ensure include_dirs for the extensions
4 ensure sources are correctly provided

MironAtHome committed:
Fix linter issue in setup.py broke the build

MironAtHome committed:
Merge master branch changes into branch with fixes

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

da-woods commented on Demos/setup.py:
```diff
@@ -8,18 +8,31 @@
+                            sources=["callback/*.pyx",
+                                     "callback/cheesefinder.c"],
+                            include_dirs=[get_include(),
+                                          "callback"]),
```
The "cheese" extension doesn't involve numpy so doesn't need to be in the `else` clause of `try: from numpy import get_include`

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome commented on Demos/libraries/setup.py:
```diff
@@ -10,23 +10,32 @@
+        Extension("call_mymath",
+                  sources=["call_mymath.pyx", "mymath.c"],
+                  include_dirs=[os.getcwd()])  # path to .h file(s)
+    ])
```
Well, it didn't work on Windows to begin with, because Windows C compiler is not relaxing to old K&R style of C inferring header from .c file.
So, please understand my mindset, as I was not just fixing, but making a call, what feature works, and needs fix, and what breaks, and so, I had to chose how to support it, while enabling build to pass without failure.
The thing about enabling library build on windows is not trivial, because inferring through the rules, what compiler is used, is not a trivial task. Where GCC is considered a nearly default feature on Linux, CL.exe is not on Windows, and so, to put a line cl.exe mymath.c in else for platform 'nt' is not something I felt comfortable, and chasing Windows C compiler I just finished after learning Distutils CCompiler class with its c= CCompiler.new_compiler('nt') followed by cc.cc call, basically, emulating how platform infers where to find C compiler on Windows OS. And given the possibility of portable MingW and SygWin makes for an overly complex for a demo case.
Plus, I didn't see a harm in introducing a cython feature, to showcase, how to correctly produce Python natively compiled module, by adding a few steps to cythonize call.
I will be happy to document regarding Windows omitting to provide a demo for Library, for the above reason, and provide example codes for all of the noted cases, including MingW and the rest, only ask it to be part of separate PR ( if this sounds reasonable ). And yeah, totally omitting creating module on windows for this demo case would require additional changes to the script actually executing the code, the run*.py script. Which does seem like an overkill in terms of conditionals.

MironAtHome committed:
Polish proposed changes.
- reduce the number of parameters in setup.py to just necessary
- remove superfluous call to get_include()
- remove specific directory to search for include files
- replace explicit module name(s) with *, as those match .pyx file name
- remove DLL_EXPORT directive from custom .c files in cheese and my_mathcall modules

MironAtHome commented on Demos/callback/cheesefinder.c:
```diff
@@ -11,7 +11,7 @@ static char *cheeses[] = {
 };
 
-void find_cheeses(cheesefunc user_func, void *user_data) {
+DLL_EXPORT void find_cheeses(cheesefunc user_func, void *user_data) {
```
Please review changes, as per above.

MironAtHome commented:
Please review changes, as per above.
Most requests fixed.
Asking to document omission with regards to Windows for call_mymath into a separate PR ( to follow shortly ).
A through documentation would require covering 2 additional compilers, mingw and cygwin.
Potentially, I would like to add OneAPI Intel compiler into the fold, as it has some very nice features, so, it looks like a very - very nice opportunity to add this compiler to the fold of tools on the list of possible choices. In an ideal world to run a small code performance comparison. I do like racing :)

da-woods commented:
> A through documentation would require covering 2 additional compilers, mingw and cygwin.

It's very difficult to use anything except MSVC on Windows.  The most promising open-source option is clang-cl. This would probably work perfectly except that setuptools doesn't recognise it.

I really would not try to support or document mingw or cygwin.

da-woods commented on Demos/setup.py:
```diff
@@ -8,18 +8,31 @@
+                            sources=["libraries/*.pyx",
+                                     "libraries/mymath.c"],
+                            include_dirs=[get_include(),
+                                          "libraries"]),
```
I think I didn't make this clear.

This file should just exclude the libraries and callback directories not attempt to build them. If people want to build the libraries or callback examples then they should use the setup files in those directories.

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Further improve quality of the Demos code
for callback and libraries .h file replace custom define with Python standard exports.h define Py_EXPORTED_SYMBOL
for setup.py script under libraries module improve building library using distutil compile, create_static_lib ccompiler class methods for Microsoft Windows platform

MironAtHome committed:
Correction to libraries Demo for linux variant.
Add python's include directory on command line to ensure correct build

MironAtHome committed:
Further improve build of cython Demo modules
Modify local setup.py under ./Demos/libraries to pick up its name from PYX file
Modify common setup.py under ./Demos so that no dependency on numpy to build callback
Retain common setup.py stopping on error immediatelly

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'miron/fix-demo-modules' of https://github.com/MironAtHome/cython into miron/fix-demo-modules

scoder commented on Demos/libraries/setup.py:
```diff
@@ -2,30 +2,47 @@
-              library_dirs=[os.getcwd()],  # path to .a or .so file(s)
-              libraries=['mymath'])
-])
+if os.path.exists("libmymath.a") or os.path.exists("mymath.lib"):
```
We call `sys.exit()` above if these files do not exist, so you don't need to make the following code conditional.

scoder commented on Demos/libraries/setup.py:
```diff
@@ -2,30 +2,47 @@
+        print(ex)
+        if not os.path.exists("libmymath.a"):
+            print("Error building external library, please create libmymath.a manually.")
+            sys.exit(1)
```
It looks like this part can be pulled out of the build cases and done generally, in the end. (I.e. do "try … if … elif" instead of "if … try … elif … try".)

MironAtHome replied:
Variables applied to handle platform specific library name.
Added *.split(os.path.pathsep) to make directory list parameter portable
Hopefully this closes the gap

scoder reviewed (commented):
I wonder if it's really worth showing two ways to build the libraries. If setuptools can do it, that's probably enough. The intention here is not to show how to build an external library because users who want to integrate with an external library probably have it available or at least know how to build it.

scoder commented on Demos/setup.py:
```diff
@@ -1,25 +1,61 @@
 
-ext_modules = cythonize("**/*.pyx", exclude="numpy_*.pyx")
+ext_modules = cythonize("**/*.pyx",
+                            exclude=["numpy_*.pyx",
```
It's not a big file, nor an important one. Just run a code formatter over it.

scoder commented on Demos/callback/cheesefinder.h:
```diff
@@ -1,2 +1,4 @@
-typedef void (*cheesefunc)(char *name, void *user_data);
-void find_cheeses(cheesefunc user_func, void *user_data);
+#include <exports.h>
```
> it is considered a good taste to use, as much as possible, platform features, and add new features only when those are really needed.

I agree in general. However, this is example code that should resemble a real-world scenario. And that scenario is unlikely to involve a library that uses Python headers and definitions, because the goal is to write code that newly connects Python to an existing C library. Thus my line of thought above. Please read it and tell me your opinion about it.

MironAtHome replied:
Replaced is the exports.h with Python.h.

MironAtHome committed:
Per code review request
modifying #include directive from <exports.h> to <Python.h> to deduce chance of name collision

MironAtHome committed:
Further improve setup.py for Demos and under libraries
Use variable for library name
Ensure plaform specific handling of include directories
Fix header line of setup.py under ./libraries
Apply python formatter to align code

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

scoder commented on Demos/libraries/setup.py:
```diff
@@ -1,33 +1,60 @@
+                f"gcc -I\"{"\" -I\"".join(py_include.split(os.path.pathsep))}\" -shared -fPIC -c {c_sources} -o mymath.o"
+            )
+            == 0
+        )
```
This is barely readable to me – way too many special characters and escapes. And it hides an important call (with side effects) inside of an `assert` statement, which may not get executed in some configurations. (I'm aware that it was like this before, but now is the best time to clean it up.) We should also use `subprocess` instead of `os.system`.

```suggestion
            c_includes = [f'-I{path}' for path in py_include.split(os.path.pathsep)]
            subprocess.check_call(["gcc", "-shared", "-fPIC", "-c"] + c_includes + [c_sources, "-o", "mymath.o"])
```

By its name, `c_sources` should rather be a list of strings than a string. That would also prevent (hypothetical) issues with spaces in file paths etc.

The same applies to the main `Demos/setup.py` script.

MironAtHome replied:
We can't assume that on Windows include directories will not have spaces, hence, need to wrap in quotes.
So, some hairiness may be a bit of a tax on the eyesight, but does help to ensure error free execution.
Thanks for the suggestion regarding subprocess. I will take a look if I can do a jump to setuptools at once, or use this suggestion, since it does look more like a change, over iterative fix.

scoder replied:
> We can't assume that on Windows include directories will not have spaces, hence, need to wrap in quotes.

I don't think we need to do that if we pass the arguments separately (as in the `subprocess` call that I suggested), rather than pushing them through a command line string.

scoder commented on Demos/libraries/mymath.h:
```diff
@@ -1 +1,3 @@
-double sinc(double);
+#include <Python.h>
+
+Py_EXPORTED_SYMBOL double sinc(double);
```
I still think that a C library example should not be using `Python.h` and `Py_EXPORTED_SYMBOL`. It's a C library, not a Python extension.

MironAtHome replied:
So, you mean to say you have a very strong emphasis on how to craft external C libraries, without Python provisions?
I understand, it just looked to me that a gentle suggestion to study what is available in Python ( and seeing how it supports portable code ) would help.
And it does look aesthetically pleasing to an eye a bit better than lines of #define(s) I had earlier.

scoder replied:
> you have a very strong emphasis on how to craft external C libraries

No, I just think we shouldn't give wrong impressions in our documentation.

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'master' into miron/fix-demo-modules

MironAtHome committed:
Merge branch 'cython:master' into miron/fix-demo-modules
